[threaded-animations] add testing support for remote layer tree animations

graouts opened the pull request:
#### 8740ea8994be7fb3f7b14502e4ba9b61ef9479f6
<pre>
[threaded-animations] add testing support for remote layer tree animations
<a href="https://bugs.webkit.org/show_bug.cgi?id=302077">https://bugs.webkit.org/show_bug.cgi?id=302077</a>
<a href="https://rdar.apple.com/164164711">rdar://164164711</a>

Reviewed by Simon Fraser.

In order to be able to test the state of animations associated with remote layer
tree nodes, we add a new `UIScriptController.animationStackForLayerWithID()` function
which we wrap as `UIHelper.remoteAnimationStackForElement()` in JavaScript.

To that end we add a suite of methods to convert objects related to remote layer
tree animations to JSON representations encoded as a string for transport.

Note that we haven&apos;t yet added support for producing JSON representations for filters
and offset paths. Those will be implemented as necessary as further tests are developed.

* LayoutTests/resources/ui-helper.js:
(window.UIHelper.remoteAnimationStackForElement):
* LayoutTests/webanimations/threaded-animations/scroll-driven-animations-expected.txt:
* LayoutTests/webanimations/threaded-animations/scroll-driven-animations.html:
* Source/WebCore/platform/animation/TimingFunction.h:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _animationStackForLayerWithID:]):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimation.cpp:
(WebKit::RemoteAnimation::toJSONForTesting const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimation.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationStack.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationStack.mm:
(WebKit::RemoteAnimationStack::toJSONForTesting const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationTimeline.cpp:
(WebKit::RemoteAnimationTimeline::RemoteAnimationTimeline):
(WebKit::RemoteAnimationTimeline::toJSONForTesting const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationTimeline.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationUtilities.cpp: Added.
(WebKit::toStringForTesting):
(WebKit::toJSONForTesting):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationUtilities.h: Added.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::animationStackForNodeWithIDForTesting const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::animationStackForNodeWithIDForTesting const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.mm:
(WebKit::RemoteLayerTreeNode::setAcceleratedEffectsAndBaseValues):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
(WebKit::RemoteScrollingCoordinatorProxy::animationStackForNodeWithIDForTesting const):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.mm:
(WebKit::RemoteLayerTreeEventDispatcher::animationStackForNodeWithIDForTesting const):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
(WebKit::RemoteScrollingCoordinatorProxyMac::animationStackForNodeWithIDForTesting const):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
(WTR::UIScriptController::animationStackForLayerWithID const):
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.h:
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::UIScriptControllerCocoa::animationStackForLayerWithID const):

Canonical link: <a href="https://commits.webkit.org/302673@main">https://commits.webkit.org/302673@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/93513a2d0380f4f26f570012f4fa97a85a2bf2c2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129852 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2113 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40709 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137244 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/81335 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/2f0e3d6f-f813-4907-a5a8-5a7985bfe6ee) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131723 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2066 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2003 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98921 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/66731 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/84a80255-0874-4606-9095-a16e1e64950e) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132799 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/1571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116299 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79606 "Passed tests") | | [❌ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b4d4dfe8-0246-4dbc-a4ad-2fa090097f46) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/1482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34431 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80515 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109983 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34933 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139724 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1907 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/1783 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/107425 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1952 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112646 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107306 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1543 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31129 "Passed tests") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54709 "Hash 93513a2d for PR 53521 does not build (failure)") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20256 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1980 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65349 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1794 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1829 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1903 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
```diff
@@ -328,6 +328,14 @@
+
+const RemoteAnimationStack* RemoteScrollingCoordinatorProxyMac::animationStackForNodeWithIDForTesting(WebCore::PlatformLayerIdentifier layerID) const
+{
+    m_eventDispatcher->lockForAnimationChanges();
+    auto* animationStack = m_eventDispatcher->animationStackForNodeWithIDForTesting(layerID);
+    m_eventDispatcher->unlockForAnimationChanges();
+    return animationStack;
```
Not super confident about this code here. Is this right @mattwoodrow?

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimation.mm:
```diff
@@ -0,0 +1,91 @@
+NSDictionary *RemoteAnimation::toJSONForTesting() const
+{
+    auto convertProperties = [](const OptionSet<WebCore::AcceleratedEffectProperty>& properties) {
+        NSMutableArray *convertedProperties = [[NSMutableArray alloc] init];
```
This is a leak. You need to `adoptNS()` and use a `RetainPtr`.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimation.mm:
```diff
@@ -0,0 +1,91 @@
+    };
+
+    auto convertKeyframes = [](const Vector<WebCore::AcceleratedEffect::Keyframe>& keyframes) {
+        NSMutableArray *convertedKeyframes = [[NSMutableArray alloc] init];
```
Ditto

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimation.mm:
```diff
@@ -0,0 +1,91 @@
+        NSMutableArray *convertedKeyframes = [[NSMutableArray alloc] init];
+        for (auto& keyframe : keyframes) {
+            NSDictionary *convertedValues = WebKit::toJSONForTesting(keyframe.values(), keyframe.animatedProperties());
+            NSMutableDictionary *convertedKeyframe = [convertedValues mutableCopy];
```
This is a leak too.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationStack.mm:
```diff
@@ -214,6 +215,22 @@
 #endif
 }
 
+NSDictionary *RemoteAnimationStack::toJSONForTesting() const
```
This remote animation code doesn't have to be Apple-platform specific (it's not in an Apple platform directory), and it's not other than this ObjC testing code. Consider not using ObjC types for the JSON stuff.

graouts replied:
Agreed, I've switched over to using `JSON::Value` and `JSON::Object` throughout the updated version of the patch.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimation.cpp:
```diff
@@ -50,6 +51,40 @@ void RemoteAnimation::apply(WebCore::AcceleratedEffectValues& values)
+Ref<JSON::Object> RemoteAnimation::toJSONForTesting() const
+{
+    auto convertProperties = [](const OptionSet<WebCore::AcceleratedEffectProperty>& properties) {
+        auto convertedProperties = JSON::Array::create();
```
We prefer putting `Ref` on the left side instead of just `auto`.

smfr commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
```diff
@@ -1170,6 +1174,25 @@ - (NSString *)_webContentProcessVariantForFrame:(_WKFrameHandle *)frameHandle
+        if (!layerID)
+            return nullptr;
+        WebCore::PlatformLayerIdentifier platformLayerID { ObjectIdentifier<WebCore::PlatformLayerIdentifierType>(layerID), _page->legacyMainFrameProcess().coreProcessIdentifier() };
+        if (auto nodeStack = downcast<WebKit::RemoteLayerTreeDrawingAreaProxy>(_page->protectedDrawingArea())->animationStackForNodeWithIDForTesting(platformLayerID))
```
`RefPtr nodeStack`

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationStack.mm:
```diff
@@ -214,6 +215,22 @@
 
+Ref<JSON::Object> RemoteAnimationStack::toJSONForTesting() const
+{
+    auto convertedAnimations = JSON::Array::create();
```
`Ref`

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationStack.mm:
```diff
@@ -214,6 +215,22 @@
+        convertedAnimations->pushObject(animation->toJSONForTesting());
+    }
+
+    auto object = JSON::Object::create();
```
`Ref`

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationTimeline.cpp:
```diff
@@ -40,6 +41,15 @@ RemoteAnimationTimeline::RemoteAnimationTimeline(TimelineID identifier, std::opt
 
+Ref<JSON::Object> RemoteAnimationTimeline::toJSONForTesting() const
+{
+    auto object = JSON::Object::create();
```
`Ref`

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationUtilities.cpp:
```diff
@@ -0,0 +1,305 @@
+        return object;
+    };
+
+    auto convertedValues = JSON::Object::create();
```
`Ref`

graouts committed:
[threaded-animations] add testing support for remote layer tree animations
https://bugs.webkit.org/show_bug.cgi?id=302077
rdar://164164711

Reviewed by Simon Fraser.

In order to be able to test the state of animations associated with remote layer
tree nodes, we add a new `UIScriptController.animationStackForLayerWithID()` function
which we wrap as `UIHelper.remoteAnimationStackForElement()` in JavaScript.

To that end we add a suite of methods to convert objects related to remote layer
tree animations to JSON representations encoded as a string for transport.

Note that we haven't yet added support for producing JSON representations for filters
and offset paths. Those will be implemented as necessary as further tests are developed.

* LayoutTests/resources/ui-helper.js:
(window.UIHelper.remoteAnimationStackForElement):
* LayoutTests/webanimations/threaded-animations/scroll-driven-animations-expected.txt:
* LayoutTests/webanimations/threaded-animations/scroll-driven-animations.html:
* Source/WebCore/platform/animation/TimingFunction.h:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _animationStackForLayerWithID:]):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimation.cpp:
(WebKit::RemoteAnimation::toJSONForTesting const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimation.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationStack.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationStack.mm:
(WebKit::RemoteAnimationStack::toJSONForTesting const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationTimeline.cpp:
(WebKit::RemoteAnimationTimeline::RemoteAnimationTimeline):
(WebKit::RemoteAnimationTimeline::toJSONForTesting const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationTimeline.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationUtilities.cpp: Added.
(WebKit::toStringForTesting):
(WebKit::toJSONForTesting):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationUtilities.h: Added.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::animationStackForNodeWithIDForTesting const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::animationStackForNodeWithIDForTesting const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.mm:
(WebKit::RemoteLayerTreeNode::setAcceleratedEffectsAndBaseValues):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
(WebKit::RemoteScrollingCoordinatorProxy::animationStackForNodeWithIDForTesting const):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.mm:
(WebKit::RemoteLayerTreeEventDispatcher::animationStackForNodeWithIDForTesting const):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
(WebKit::RemoteScrollingCoordinatorProxyMac::animationStackForNodeWithIDForTesting const):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
(WTR::UIScriptController::animationStackForLayerWithID const):
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.h:
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::UIScriptControllerCocoa::animationStackForLayerWithID const):

Canonical link: https://commits.webkit.org/302673@main
